### DISCUSSION TITLE
msvc libarchive 同时存在 archive 和 archive_static.lib，默认为static，但是编译后还是需要 archive.dll

### DISCUSSION THREAD

**Discussion by kezhengjie** [xmake-io/xmake#3931] [Q&A] [answered]
作者您好！
platform: windows
我这边  add_requires("libarchive") 后
为我的target  add_packages("libarchive")
看了下配置默认是 shared: false
![image](https://github.com/xmake-io/xmake/assets/55621471/596be2e6-b918-4658-8b82-c1bd77ae8626)
但是我看生成的有 archive.lib 和 archive_static.lib
![image](https://github.com/xmake-io/xmake/assets/55621471/76453004-a0e7-42f1-883b-471ab363e3e8)
实际编译后运行时也是需要 archive.dll ，感觉是 archive.lib 和 archive_static.lib 同时link了
请问是我的用法有问题吗，或者有没有办法只 link archive_static.lib ，因为我写的这个程序需要尽可能少dll所以有这疑惑。
期待您的回复。

**Comment by waruqi** [comment]
这种一般是包没做好，内部 on_install 逻辑没区分 static/shared 。。同时安装了。。你可以提 pr 改进这个包，通过 package:config("shared") 区分下 static/shared 。。分别独立安装，就没事了。。可以参考其他的包配置

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake-repo/blob/051ca0fd8fa1826713dd3ec7b8b9b9ea8190232b/packages/l/libarchive/xmake.lua#L16 

我看这个包有做区分，但是应该是这里没生效。。具体你得分析下这个包的 cmakelists.txt 配置。看下为什么没生效，然后提个 pr 过来改进下

**Reply by waruqi** [reply]
我猜它这个包的 cmakelists 压根没做 static/shared 处理，总是会一次性生成所有。。

这种，可以走 os.rm 来区分，参考：https://github.com/xmake-io/xmake-repo/blob/051ca0fd8fa1826713dd3ec7b8b9b9ea8190232b/packages/m/mysql/xmake.lua#L91

**Reply by kezhengjie** [reply]
好的，谢谢。我这边详细看看

**Reply by kezhengjie** [reply]
我这边看了下，他的CMakeLists.txt确实是同时生成 archive 和 archive_static 的 
![image](https://github.com/xmake-io/xmake/assets/55621471/6ae1945a-e98b-43e8-a332-4faa7a6faa88)

我这边也尝试修改了 xmake.lua ，根据shared删除了 archive.lib 和 archive.dll
![image](https://github.com/xmake-io/xmake/assets/55621471/15b918e0-614d-4b94-b67b-1cd9584884a7)

可是install会失败，感觉是我删除了 archive.lib 导致的。
![image](https://github.com/xmake-io/xmake/assets/55621471/4aa1f40b-38e7-4b4d-b031-961faf8db4f1)

我原以为到生成 lib 和 dll 这个阶段就结束了，看起来还有些其他步骤，请问这个我该如何进一步定位问题呢？

**Reply by kezhengjie** [reply]
我知道了 应该是因为到这个 on_test 
![image](https://github.com/xmake-io/xmake/assets/55621471/b89b4b4a-f540-4493-810f-08ff9910e54c)
这个test应该是用shared的那个lib跑的test，请问test这部分该怎么修改合适呢？

**Reply by waruqi** [reply]
库应该对了，但是 on_test 以及用户端，引入这个 api 可能默认当成了 shared

到它们头文件里面找下，应该有类似 XXX_EXPORT 的宏定义，来切的。。包配置 on_load 里面根据 shared/static 。。用 `package:add("defines", "XXX_EXPORT")` 定义上相关宏

**Reply by kezhengjie** [reply]
我这边找到了 dllexport 的地方
![image](https://github.com/xmake-io/xmake/assets/55621471/01411b11-c47f-4862-bd8a-3720978a6ff6)

在on_load 中进行了 定义
![image](https://github.com/xmake-io/xmake/assets/55621471/55346b93-ec1e-4b19-af20-40852eb52703)

可是在 on_test 阶段报错，这次看函数签名应该是非 shared 的
![image](https://github.com/xmake-io/xmake/assets/55621471/56703a42-fd9f-48bd-9a75-975b7b641972)

我还尝试了在  on_load 中
pacakge:add(links","archive_static") 
虽然感觉不对，这个应该是自动链接的，也还是不行。

奇怪的是如果我注释掉 on_test ， 我的 target 是能正常编译并链接 archive_static 的，运行起来也一切正常，
请问这该如何排查呢？

xmake.lua 修改部分
![image](https://github.com/xmake-io/xmake/assets/55621471/505b851c-0ef0-4600-a0bd-c4d80ef33e79)

**Reply by waruqi** [reply, marked as answer]
要根据 `if package:is_plat("windows")  and not package:config("shared") then` 仅对 window static 添加

**Reply by kezhengjie** [reply]
可以了，一切正常，谢谢作者~
我去提交一下pr
我一直以为这个 on_test 是根据现有的编译环境生成对象来校验，这样看起来还有很多细节？

**Reply by waruqi** [reply]
on_test 模拟的是用户工程的包配置环境
